### ISSUE TITLE
wrong suggestion for `needless_range_loop` on mutating iterator

### ISSUE THREAD

**Issue by palango** [rust-lang-nursery/rust-clippy#2068]
I get a false report about `needless_range_loop` with the following code.

```
impl BitXor for NodeId {
    type Output = NodeId;

    fn bitxor(self, rhs: NodeId) -> Self::Output {
        let mut b = [0u8; LEN];
        for i in 0..LEN {
            b[i] = self.data[i] ^ rhs.data[i];
        }

        NodeId { data: b }
    }
}
```

```
warning: the loop variable `i` is used to index `b`
  --> src/main.rs:67:9
   |
67 | /         for i in 0..LEN {
68 | |             b[i] = self.data[i] ^ rhs.data[i];
69 | |         }
   | |_________^
   |
   = note: #[warn(needless_range_loop)] on by default
   = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.161/index.html#needless_range_loop
help: consider using an iterator
   |
67 |         for (i, <item>) in b.iter().enumerate().take(LEN) {
   |             ^^^^^^^^^^^
```

**Comment by clarcharr** [comment]
Personally I'd suggest:

```
for (i, dest) in self.data.iter_mut().enumerate() {
    *dest = self.data[i] ^ rhs.data[i];
}
```

**Comment by josh0588** [comment]
Hi I'd like to start contributing and take this on as my first bug

**Comment by bluss** [comment]
Can't zip be used here? nested zip should still use the specialization that makes it bounds check free (and index free).

It's not very pretty:

```rust
for ((dest, &x), &y) in b.iter_mut().zip(&self.data).zip(&rhs.data) {
    *dest = x ^ y;
}
```

The crucial drawback is the loss of the "error handling" (= panic) if the inputs self.data and rhs.data are shorter than b.

**Comment by mcarton** [comment]
Of course itertools has the solution for that [`itertools::zip_eq`](https://docs.rs/itertools/0.6.5/itertools/fn.zip_eq.html).

**Comment by bofh69** [comment]
I just tried it in the playground:
[Playground link](https://play.rust-lang.org/?gist=9566301a192832aaea5109d1876c90ab&version=stable&mode=debug&edition=2015)

The message is now suggesting iter_mut:
```
warning: the loop variable `i` is used to index `b`
  --> src/main.rs:11:18
   |
11 |         for i in 0..LEN {
   |                  ^^^^^^
   |
   = note: #[warn(clippy::needless_range_loop)] on by default
   = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.212/index.html#needless_range_loop
help: consider using an iterator
   |
11 |         for (i, <item>) in b.iter_mut().enumerate().take(LEN) {
   |             ^^^^^^^^^^^    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
